code/__HELPERS/_logging.dm

```diff
@@ -125,58 +128,87 @@
 
 	return english_list(comps, nothing_text="0", and_text="|", comma_text="|")
 
-//more or less a logging utility
-/proc/key_name(var/whom, var/include_link = null, var/include_name = 1, var/highlight_special_characters = 1)
+/* Helper procs for building detailed log lines */
+/proc/key_name(whom, include_link = null, include_name = TRUE, highlight_special_characters = TRUE)
 	var/mob/M
 	var/client/C
 	var/key
+	var/ckey
+	var/fallback_name
 
-	if(!whom)	return "*null*"
+	if(!whom)
+		return "*null*"
 	if(istype(whom, /client))
 		C = whom
 		M = C.mob
 		key = C.key
+		ckey = C.ckey
 	else if(ismob(whom))
 		M = whom
 		C = M.client
 		key = M.key
+		ckey = M.ckey
+	else if(istext(whom))
+		key = whom
+		ckey = ckey(whom)
+		// C = GLOB.directory[ckey]
+		// if(C)
+		// 	M = C.mob
 	else if(istype(whom, /datum/mind))
-		var/datum/mind/D = whom
-		key = D.key
-		M = D.current
-		if(D.current)
-			C = D.current.client
-	else if(istype(whom, /datum))
-		var/datum/D = whom
-		return "*invalid:[D.type]*"
-	else
-		return "*invalid*"
+		var/datum/mind/mind = whom
+		key = mind.key
+		ckey = ckey(key)
+		if(mind.current)
+			M = mind.current
+			if(M.client)
+				C = M.client
+		else
+			fallback_name = mind.name
+	else // Catch-all cases if none of the types above match
+		var/swhom = null
+
+		if(istype(whom, /atom))
+			var/atom/A = whom
+			swhom = "[A.name]"
+		else if(istype(whom, /datum))
+			swhom = "[whom]"
+
+		if(!swhom)
+			swhom = "*invalid*"
+
+		return "\[[swhom]\]"
 
 	. = ""
 
-	if(key)
-		if(include_link && C)
-			. += "<a href='?priv_msg=\ref[C]'>"
+	if(!ckey)
+		include_link = FALSE
 
-		if(C && C.holder && C.holder.fakekey && !include_name)
+	if(key)
+		if(C?.holder && C.holder.fakekey && !include_name)
+			if(include_link)
+				. += "<a href='?priv_msg=[REF(C)]'>"
 			. += "Administrator"
 		else
+			if(include_link)
+				. += "<a href='?priv_msg=[REF(ckey)]'>"
 			. += key
```

Comment:
Ckey refs dont work with our admin system.

Reply:
client procs handle translating this